### PR TITLE
Improve performance when used with RPG control stack

### PR DESCRIPTION
**Pull request by kelia**
Sorry for the large PR. 

- [x] Specify topic to visualize MPC path (useful when multiple MPC nodes are running)
- [x] Solve MPC from hover conditions after OFF command has been received.

### CODE REVIEW

File: src/mpc_controller.cpp [hunk 1]
```diff
@@ -28,70 +28,72 @@
...
-  timing_preparation_(T(1e-3)),
-  est_state_((Eigen::Matrix<T, kStateSize, 1>() <<
-    0, 0, 0, 1, 0, 0, 0, 0, 0, 0).finished()),
-  reference_states_(Eigen::Matrix<T, kStateSize, kSamples+1>::Zero()),
-  reference_inputs_(Eigen::Matrix<T, kInputSize, kSamples+1>::Zero()),
-  predicted_states_(Eigen::Matrix<T, kStateSize, kSamples+1>::Zero()),
-  predicted_inputs_(Eigen::Matrix<T, kInputSize, kSamples>::Zero()),
-  point_of_interest_(Eigen::Matrix<T, 3, 1>::Zero())
-{
+    const ros::NodeHandle& nh, const ros::NodeHandle& pnh, const std::string& topic) :
```
Comment (foehnx) [inline src/mpc_controller.cpp:33]: why four spaces indent?

File: include/rpg_mpc/mpc_controller.h [hunk 2]
```diff
@@ -74,8 +75,10 @@ class MpcController {
   static_assert(kInputSize == 4,
     "MpcController: Wrong model size. Number of inputs does not match.");
 
-  MpcController(const ros::NodeHandle & nh, const ros::NodeHandle & pnh);
-  MpcController() : MpcController(ros::NodeHandle(), ros::NodeHandle("~")) {}
+  MpcController(const ros::NodeHandle & nh, const ros::NodeHandle & pnh, const std::string& topic);
```
Comment (foehnx) [inline include/rpg_mpc/mpc_controller.h]: rather spare writing constructors with multiple definitions, just give it default arguments like
```
MpcController(
  const ros::NodeHandle & nh,
  const ros::NodeHandle & pnh,
  const std::string& topic = "mpc/trajectory_predicted");
```

File: src/mpc_controller.cpp [hunk 3]
```diff
@@ -106,7 +108,13 @@ quadrotor_common::ControlCommand MpcController<T>::run(
 
   // Get the feedback from MPC.
   mpc_wrapper_.setTrajectory(reference_states_, reference_inputs_);
-  mpc_wrapper_.update(est_state_, do_preparation_step);
+  if (solve_from_scratch_) {
+    ROS_INFO("Solving MPC from Hover Conditions.");
```
Comment (foehnx) [inline src/mpc_controller.cpp]: This message is missleading.
It only uses hover as initial guess...
additionally, have you tried doing more than one solving step after resetting?
is this tested?

File: src/mpc_controller.cpp [hunk 4]
```diff
@@ -161,80 +167,74 @@ bool MpcController<T>::setReference(
...
-    std::list<quadrotor_common::TrajectoryPoint>::const_iterator iterator(
-      reference_trajectory.points.begin());
-    for(int i=0; i<kSamples+1; i++)
-    {
-      while(iterator->time_from_start.toSec() < i*dt &&
-            iterator!=reference_trajectory.points.end())
-      {
+        reference_trajectory.points.front().heading,
+        Eigen::Matrix<T, 3, 1>::UnitZ()));
+    q_orientation = reference_trajectory.points.front().orientation.template cast<T>() * q_heading;
```
Comment (foehnx) [inline src/mpc_controller.cpp:174]: Have you verified that this change in rotation sequence is correct?

**Review by foehnx — APPROVED** [review]
The code style changes make lines go well beyond 80 cols, even 100 cols.
would be nice if this is reverted.
is this all tested?
even on real platform?

File: src/mpc_controller.cpp (hunk 1 above)
Comment (kelia) [reply src/mpc_controller.cpp:33]: I used 4 spaces indent for function parameters on newlines. Makes it easier to discriminate between function header and body, which has two spaces indent.

File: include/rpg_mpc/mpc_controller.h [hunk 5]
```diff
@@ -57,60 +57,60 @@ enum STATE {
...
+                "MpcController: Wrong model size. Number of states does not match.");
   static_assert(kInputSize == 4,
-    "MpcController: Wrong model size. Number of inputs does not match.");
+                "MpcController: Wrong model size. Number of inputs does not match.");
 
-  MpcController(const ros::NodeHandle & nh, const ros::NodeHandle & pnh, const std::string& topic);
-  MpcController() : MpcController(ros::NodeHandle(), ros::NodeHandle("~"), "mpc/trajectory_predicted") {}
-  MpcController(const ros::NodeHandle & nh, const ros::NodeHandle & pnh) : MpcController(nh, pnh, "mpc/trajectory_predicted") {}
-  MpcController(const std::string& topic) : MpcController(ros::NodeHandle(), ros::NodeHandle("~"), topic) {}
+  MpcController(const ros::NodeHandle& nh,
```
Comment (foehnx) [inline include/rpg_mpc/mpc_controller.h:78]: nice!

**Review by foehnx — APPROVED** [review]

**Pull request merged**